cupy/core/_routines_math.pyx

```diff
@@ -216,32 +372,126 @@ cdef ndarray scan(ndarray a, ndarray out=None):
+        kern_add = _add_scan_blocked_sum_kernel(out.dtype, op)
         kern_add(grid=((a.size - 1) // (2 * block_size),),
                  block=(2 * block_size - 1,),
                  args=(out,))
     return out
 
 
+cdef ndarray _batch_scan_op(ndarray a, op, dtype=None, out=None):
+    batch_size = a.shape[1]
+    # TODO(ecastill) replace this with "_reduction._block_size" once it is
+    # properly exposed
+    block_size = 512
+    # Since we need to pad each batch we spawn more threads as some
+    # of them will be idle
+    # Calc the total number of blocks
+    padded_bs = 1 << ((batch_size - 1).bit_length())
+    if padded_bs > block_size:
+        blocks_per_batch = (batch_size - 1) // block_size + 1
+        padded_bs = block_size * blocks_per_batch
+    padded_size = a.size // batch_size * padded_bs
+
+    kern_scan = _inclusive_batch_scan_kernel(a.dtype, block_size, op)
+    kern_scan(grid=((padded_size - 1) // (block_size) + 1,),
+              block=(block_size,),
+              args=(a, out, batch_size),
+              shared_mem=a.itemsize * block_size)
+    if batch_size > block_size:
+        blocked_sum = out[:, block_size-1::block_size]
+        _batch_scan_op(blocked_sum, op, dtype, blocked_sum)
+        kern_add = _add_scan_batch_blocked_sum_kernel(
+            out.dtype, op, block_size)
+        kern_add(
+            grid=((out.size - 1) // (block_size) + 1,),
+            block=(block_size,),
+            args=(out, batch_size))
+    return out
+
+
+cdef _axis_to_first(ndarray x, axis):
+    trans = [axis] + [a for a in range(x.ndim) if a != axis]
+    pre = list(range(1, axis + 1))
+    succ = list(range(axis + 1, x.ndim))
+    revert = pre + [0] + succ
+    return trans, revert
+
+
+cdef _proc_as_batch(ndarray x, axis, dtype, op):
+    if x.shape[axis] == 0:
+        return cupy.empty_like(x)
+    trans, revert = _axis_to_first(x, axis)
+    t = x.transpose(trans)
+    s = t.shape
+    r = t.reshape(x.shape[axis], -1)
+    # This is to use the current implemented fast scan
+    # TODO(ecastill) merge with above transformations
+    r = r.transpose((1, 0))
+    r = _batch_scan_op(r, op, dtype, r)
+    r = r.transpose((1, 0))
+    return r.reshape(s).transpose(revert)
+
+
+cpdef scan_core(ndarray a, axis, op, dtype=None, ndarray out=None):
+    if out is None:
+        if dtype is None:
+            kind = a.dtype.kind
+            if kind == 'b':
+                dtype = numpy.dtype('l')
+            elif kind == 'i' and a.dtype.itemsize < numpy.dtype('l').itemsize:
+                dtype = numpy.dtype('l')
+            elif kind == 'u' and a.dtype.itemsize < numpy.dtype('L').itemsize:
+                dtype = numpy.dtype('L')
+            else:
+                dtype = a.dtype
+        result = a.astype(dtype=dtype)
+    else:
+        if (out.flags.c_contiguous or out.flags.f_contiguous):
+            result = out
+        else:
+            result = a.astype(out.dtype)
```

Comment:
The interplay between `dtype` and `out` is almost correct (according to https://github.com/cupy/cupy/pull/2919#issuecomment-575705775), except that `dtype` is ignored during calculation if `out` is not `None`. But I think it is OK since the output dype is set to `out.dtype`, which is all end users care about. I don't think it'd cause any noticeable difference.

Reply:
cf: #2903